parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java

```diff
@@ -136,6 +154,19 @@ public static Metrics footerMetrics(ParquetMetadata metadata, MetricsConfig metr
         toBufferMap(fileSchema, lowerBounds), toBufferMap(fileSchema, upperBounds));
   }
 
+  private static MessageType getParquetType(ParquetMetadata metadata, Schema expectedSchema, NameMapping nameMapping) {
+    MessageType type = metadata.getFileMetaData().getSchema();
+    if (ParquetSchemaUtil.hasIds(type)) {
+      return expectedSchema != null ? ParquetSchemaUtil.pruneColumns(type, expectedSchema) : type;
+    } else if (nameMapping != null) {
+      MessageType typeWithIds = ParquetSchemaUtil.applyNameMapping(type, nameMapping);
```

Comment:
We have to prune columns without ids. Otherwise, `ParquetSchemaUtil.convert` will assign fallback ids to columns not present in the name mapping and they will most likely conflict with ids assigned by the name mapping.

Reply:
An alternative is to build a custom visitor that would combine logic from `ApplyNameMapping` and `PruneColumns`.

Reply:
Do we need `expectedSchema` to do this?

It seems like it would be easier if we didn't need it. What we want is to use the IDs from the mapping without assigning new IDs for columns that don't have them during conversion. So why not make a version of `convert` that doesn't assign columns? That seems much easier than doing column pruning and it doesn't require a new schema -- which introduced new methods everywhere to pass that new schema.

Reply:
Agree, let me update this.

Reply:
Actually, I remember one point I forgot. We cannot have a version of `convert` that does not assign ids as we must have an id for every column in Iceberg schema. It seems what we want is a way to prune columns without ids.

Reply:
We can also add a flag to `MessageTypeToType` to throw an exception instead of assigning an id.

Reply:
What about a version of `MessageTypeToType` that removes columns without IDs rather than assigning new ones? That would be more flexible than failing if unassigned columns are present.